Documenting plotting and sensitivity functions (#308)
esqLABS/esqlabsR: vignettes/example-visualization.Rmd

---
title: "Visualization and reporting"
output: rmarkdown::html_vignette
#output: pdf_document
vignette: >
  %\VignetteIndexEntry{Visualization and reporting}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
knitr::opts_knit$set(
  root.dir = "../tests/data/TestProject/Code/"
)
```

> ```{r, echo = FALSE, results = "hide", message = F}

Author: Some code blocks are in `echo = FALSE`. The idea is that these objects have been created in previous vignettes, so they line up in a sequence of vignettes actually.

observedData <- loadObservedData(projectConfiguration = projectConfiguration, sheets = dataSheets)
scenarioConfiguration <- ScenarioConfiguration$new(projectConfiguration)
OutputPaths <- enum(list(
  Aciclovir_PVB = "Organism|PeripheralVenousBlood|Aciclovir|Plasma (Peripheral Venous Blood)"
))
simulatedScenarios <- runScenarios(
  scenarioNames = c("TestScenario"),
  scenarioConfiguration = scenarioConfiguration,
  customParams = NULL, saveSimulationsToPKML = FALSE
)
```

### Workflow

Plotting the simulation results is an important part of model diagnostics and quality control. Simulated modeling scenarios can be passed to plotting functions from the `{ospsuite}` package to create uniformly-looking plots.

`DataCombined` is a class used to store matching observed and simulated data. Initialize a new class instance and populate it with data with the following code:

```{r datacombined}
dataCombined <- DataCombined$new()
dataCombined$addDataSets(observedData, names = "Observed", groups = "Aciclovir")
dataCombined$addSimulationResults(simulatedScenarios$TestScenario$results, names = "Simulated", groups = "Aciclovir")
dataCombined$toDataFrame()
```

> The simulation results are stored in a list returned by the `runScenarios()` function. Plotting and visualization is performed by storing these results along with matching observed data in an instance of the `DataCombined` class and passing it to plotting functions.

Reviewer: ```suggestion
The simulation results are stored in a list returned by the `runScenarios()` function. Plotting and visualization is performed by storing these results along with matching observed data in a `DataCombined` object and passing it to plotting functions.
```

> Plotting functions in the `{ospsuite}` package are wrappers around `{tlf}` plotting functions that provide default plot configuration options. All of them accept instances of `DataCombined` class as the data source. 

Reviewer: I am not sure if we need to mention anything about implementation details here; i.e. about `{tlf}`.

Author: I think it doesn't hurt, even if mentioning `{tlf}` is not directly beneficial to the users.

> Time profile plots visualize the pharmacokinetics of the drug in question and help assess if the observed data (shown with scatter plots) match the simulated data (shown with line plots).

Reviewer: ```suggestion
Time profile plots visualize the pharmacokinetics of the drug in question and help assess if the observed data (represented by points and error bars) match the simulated data (represented by lines).

```

```{r plot time profile, fig.width=7, fig.height=4}
plotIndividualTimeProfile(dataCombined)
```

> Observed vs simulated plots show if simulated time points and observed time points follow a linear trend.

Reviewer: ```suggestion
Observed versus simulated plots show if simulated time points and observed time points follow a linear trend.
```

```{r plot obs vs pred, fig.width=7, fig.height=4}
plotObservedVsSimulated(dataCombined)
```

Residual plots show if there is a systematic bias in how the simulation represents values either in high-concentration or low-concentration regions, or, alternatively, in early or late time periods.

```{r residuals vs simulated, fig.width=7, fig.height=4}
plotResidualsVsSimulated(dataCombined)
```

```{r residuals vs time, fig.width=7, fig.height=4}
plotResidualsVsTime(dataCombined)
```

> The resulting plots (returned by plotting functions) are instances of `ggplot` class. They can be modified and saved to files with `{ggplot2}` functions:

Reviewer: ```suggestion
The plots returned by plotting functions are `ggplot` objects. They can be modified further and saved to files with `{ggplot2}` functions:
```

> ```{r save plots}

Reviewer: ```suggestion
```{r save-plots}
```

cf. https://github.com/Open-Systems-Pharmacology/Suite/blob/develop/CODING_STANDARDS_R.md#rmarkdown

Reviewer: Please change all chunk labels accordingly.

Reviewer: This should be resolved only when all chunk labels have been renamed to follow our coding conventions.

plotObject <- plotIndividualTimeProfile(dataCombined)
ggplot2::ggsave(filename = "../Results/aciclovir_time_profile.png", plotObject, width = 8, height = 4)
```

### Troubleshooting

> * At any time, you can check the groups assigned to the datasets in the `DataCombined` class instance by examining the output of `dataCombined$toDataFrame()`.

Reviewer: ```suggestion
* At any time, you can check the groups assigned to the datasets in the `DataCombined` object by examining the output of `dataCombined$toDataFrame()`.
```
